panel/panel.py

```diff
@@ -207,107 +185,117 @@ def fill_panel(window, update=False, **panel_filter):
-    ]
-    if not dics:
-        return
-    return min(dics, key=lambda x: abs(x["start"] - colno))
+def get_next_panel_line(buf_line, buf_errors):
+    """Get panel line for next error in buffer.
 
-
-def get_next_lineno(num, interval):
-    interval = set(interval)
-    interval.discard(num)
-    interval = list(interval)
-    interval.sort()
-
-    if num < interval[0] or interval[-1] < num:
-        return interval[0]
+    If not found this means the current line is past last error's buffer line.
+    In that case return last error's panel line incremented by one, which will
+    place panel selection in empty space between buffer sections.
+    """
+    for d in buf_errors:
+        if d["line"] > buf_line:
+            panel_line = d["panel_line"]
+            break
     else:
-        i = bisect.bisect_right(interval, num)
-        neighbours = interval[i - 1:i + 1]
-        return neighbours[1]
+        panel_line = buf_errors[-1]["panel_line"] + 1
 
+    return panel_line, panel_line
 
-def change_selection(panel_lineno, full_line=False):
-    panel = get_panel(sublime.active_window())
-    if not panel:
-        return
-
-    region = sublime.Region(panel.text_point(panel_lineno - 1, -1))
-    if full_line:
-        region = panel.line(region)
 
+def update_selection(panel, region=None):
     selection = panel.sel()
     selection.clear()
-    selection.add(region)
+    if region is not None:
+        selection.add(region)
 
-    # scroll selection into panel
-    if not panel.visible_region().contains(region):
-        panel.show(region)
 
-    # simulate scrolling to enforce rerendering of panel,
-    # otherwise selection is not updated (ST core bug)
-    panel.run_command("scroll_lines")
+def get_panel_region(row, panel, is_full_line=False):
+    region = sublime.Region(panel.text_point(row, -1))
+    if is_full_line:
+        region = panel.line(region)
+    return region
 
 
 def update_panel_selection(active_view, we_count, current_pos, **kwargs):
+    """Alter panel selection according to errors belonging to current position.
+
+    If current position is between two errors, place empty panel selection on start of next error's panel line.
+
+    If current position is past last error, place empty selection on the panel line following that of last error.
+
+    """
     if current_pos == (-1, -1):
         return
 
-    full_line = False
-    view_dict = persist.errors.get_view_dict(active_view.id())
-    if not view_dict or util.is_none_or_zero(we_count):
+    buf_errors = persist.raw_errors.get(active_view.buffer_id())
+    if not buf_errors:  # None type error with default dict??
+        return
+    buf_errors = [d for d in buf_errors if "panel_line" in d]
+    if not buf_errors:
         return
 
-    (lineno, colno) = current_pos
-    line_dicts = view_dict["line_dicts"]
+    (buf_line, col) = current_pos
+    line_dicts = [d for d in buf_errors if d["line"] == buf_line]
```

Comment:
d -> error

Reply:
I prefer renaming `buf_errors` -> plain `errors` and `d` -> `e` standing for error. What do you think?

I feel comprehensions are one of few contexts in which single char variable names make sense.
Their shortness provide additional visual structure. Yet, they should be named according to the first letter of the iterable.